pad_server/src/controller.c

```diff
@@ -138,52 +167,58 @@ void *controller_run(void *arg) {
         for (;;) {
             /* Get the message header to determine what to handle */
             header_p hdr;
-            ssize_t bread = 0;
-            size_t total_read = 0;
 
-            while (total_read < sizeof(hdr)) {
-                bread = controller_recv(&controller, (char *)&hdr + total_read, sizeof(hdr) - total_read);
-                if (bread == -1) {
-                    fprintf(stderr, "Error reading message header: %s\n", strerror(errno));
-                    break;
-                } else if (bread == 0) {
-                    fprintf(stderr, "Control box disconnected.\n");
-                    break;
-                }
-                total_read += bread;
-            }
-
-            // Error happened, do a cleanup and re-initialize connection
-            if (bread <= 0) {
+            err = controller_receive(&controller, &hdr, sizeof(hdr));
+            if (err == -1) {
+                flush_recv(&controller);
+                fprintf(stderr, "Error reading message header: %s\n", strerror(errno));
+                continue;
+            } else if (err == 0) {
+                fprintf(stderr, "Control box disconnected.\n");
                 controller_disconnect(&controller);
                 break;
             }
 
-            // TODO: handle recv errors
-
             switch ((packet_type_e)hdr.type) {
             case TYPE_CNTRL:
 
                 switch ((cntrl_subtype_e)hdr.subtype) {
                 case CNTRL_ACT_ACK:
                 case CNTRL_ARM_ACK:
                     fprintf(stderr, "Unexpectedly received acknowledgement from sender.\n");
+                    flush_recv(&controller);
```

Comment:
This is where I'd probably prefer to read the size of a `CNTRL_ARM_ACK`/`CNTRL_ACT_ACK` instead of flushing, because we can avoid tossing packets this way.